### PR TITLE
[MNOE-502] Improvement on-boarding

### PR DESCRIPTION
**Pull request by rheasunshine**
Error message when adding apps during the on-boarding
Disable user from adding conflicting apps

@ouranos @alexnoox

### CODE REVIEW

File: src/app/components/mno-app-onboarding-btn/mno-app-onboarding-btn.html [hunk 1]
```diff
@@ -36,3 +36,10 @@
     {{ 'mno_enterprise.templates.components.app_install_btn.access_addon' | translate }}
   </a>
 </div>
+
+<div ng-show="!$ctrl.isLoadingAppInstances" ng-switch on="$ctrl.appInstance.status">
+  <div ng-switch-when="CONFLICT" class="message">
```
Comment (alexnoox) [inline src/app/components/mno-app-onboarding-btn/mno-app-onboarding-btn.html]: This code block seems to be the same as here: https://github.com/maestrano/mno-enterprise-angular/pull/235/files#diff-1b3a0dd242b70bf1890c57746b33e1a4R6

Won't it display the error twice?

Comment (rheasunshine) [inline src/app/components/mno-app-onboarding-btn/mno-app-onboarding-btn.html]: Yes, my mistake. Pushing now.

**Comment by alexnoox** [comment]
@ouranos Please have a look

**Comment by alexnoox** [comment]
@ouranos Please have a look

File: src/app/views/onboarding/step2.coffee [hunk 2]
```diff
@@ -18,8 +18,24 @@ angular.module 'mnoEnterpriseAngular'
...
+        app.subcategoryNames.push(appSubcategory.name)
+      )
+      app.marketplaceApps = _.filter(vm.marketplace.apps, (marketplaceApp) ->
+        marketplaceApp != app
+      )
+      _.each(app.marketplaceApps, (marketplaceApp) ->
+        _.each(marketplaceApp.subcategories, (marketplaceAppSubCategory) ->
+          _.each(app.subcategoryNames, (subcategoryName) ->
+            if !marketplaceAppSubCategory.multi_instantiable && (marketplaceAppSubCategory.name == subcategoryName)
+              marketplaceApp.conflictingApp = app
```
Comment (alexnoox) [inline src/app/views/onboarding/step2.coffee]: This section is a bit complicated and may be simplified.
Inspire from those examples:
- https://github.com/maestrano/mno-enterprise-angular/blob/1.1/src/app/components/mno-app-onboarding-btn/mno-app-onboarding-btn.coffee
- https://github.com/maestrano/mno-enterprise-angular/blob/1.1/src/app/components/mno-app-install-btn/mno-app-install-btn.coffee

Comment (alexnoox) [inline src/app/views/onboarding/step2.coffee]: I recommend that you read the lodash doc and play with it a bit, those functions are present in other languages as well :)

https://lodash.com/docs/3.10.1

**Comment by rheasunshine** [comment]
Hey @alexnoox , I made these changed. Let me know if there is anything else. ✌️

File: src/app/views/onboarding/step2.coffee [hunk 3]
```diff
@@ -18,8 +18,17 @@ angular.module 'mnoEnterpriseAngular'
 
     # Select or deselect an app
     vm.toggleApp = (app) ->
-      # User cannot add more apps
-      return if vm.maxAppsSelected && !app.checked
+      # Add conflictingApp attribute to marketplace apps
+      names = _.map(app.subcategories, 'name')
+      marketplaceApps = _.filter(vm.marketplace.apps, (marketApp) -> marketApp != app)
```
Comment (ouranos) [inline src/app/views/onboarding/step2.coffee]: I'd extract these loops in a separate method to help with readability.
I'd also try to reduce the number of iterations.

Something like:
```coffeescript
# Add conflictingApp attribute to marketplace apps
refreshConflictingApp = (app) ->
  # List of non multi_instantiable categories
  subCats = _.map(_.filter(app.subcategories, 'multi_instantiable', false), 'name')

  for marketApp in vm.marketplace.apps
    continue if marketApp == app

    if _.some(marketApp.subcategories, (subCat) -> not subCat.multi_instantiable and subCat.name in subCats)
      marketApp.conflictingApp = app
```

**Comment by alexnoox** [comment]
@rheasunshine Please apply @ouranos comments. i will do a last review after that 😄

**Comment by rheasunshine** [comment]
@alexnoox @ouranos Changes pushed. 🍎

**Pull request merged**